packages/analytics-core/src/plugins/destination.ts

```diff
@@ -78,87 +77,94 @@ export class Destination implements DestinationPlugin {
       if (context.attempts < this.config.flushMaxRetries) {
-        context.attempts += 1;
         return true;
       }
       void this.fulfillRequest([context], 500, MAX_RETRIES_EXCEEDED_MESSAGE);
       return false;
     });
   }
 
-  scheduleTryable(list: Context[], shouldAddToQueue = false) {
-    list.forEach((context) => {
-      // Only need to concat the queue for the first time
-      if (shouldAddToQueue) {
-        this.queue = this.queue.concat(context);
-      }
-      if (context.timeout === 0) {
-        this.schedule(this.config.flushIntervalMillis);
-        return;
+  increaseAttempts(list: Context[]) {
+    const updateEventSet = new Set(list.map((context) => context.event.insert_id));
+    this.queue.forEach((context) => {
+      if (updateEventSet.has(context.event.insert_id)) {
+        context.attempts += 1;
       }
-
-      setTimeout(() => {
-        context.timeout = 0;
-        this.schedule(0);
-      }, context.timeout);
     });
   }
 
-  addToQueue(...list: Context[]) {
-    const tryable = this.getTryableList(list);
-    this.scheduleTryable(tryable, true);
-    this.saveEvents();
-  }
+  sendEventsIfReady() {
+    if (this.config.offline) {
+      return;
+    }
 
-  schedule(timeout: number) {
-    if (this.scheduled || this.config.offline) {
+    if (this.queue.length >= this.config.flushQueueSize) {
+      void this.flush(true);
+    }
+
+    if (this.scheduled) {
       return;
     }
 
     this.scheduled = setTimeout(() => {
       void this.flush(true).then(() => {
         if (this.queue.length > 0) {
-          this.schedule(timeout);
+          this.sendEventsIfReady();
         }
       });
-    }, timeout);
+    }, this.config.flushIntervalMillis);
+
+    return;
   }
 
+  // flush the queue
   async flush(useRetry = false) {
     // Skip flush if offline
     if (this.config.offline) {
       this.config.loggerProvider.debug('Skipping flush while offline.');
       return;
     }
 
-    const list: Context[] = [];
-    const later: Context[] = [];
-    this.queue.forEach((context) => (context.timeout === 0 ? list.push(context) : later.push(context)));
-
     if (this.scheduled) {
       clearTimeout(this.scheduled);
       this.scheduled = null;
     }
 
-    const batches = chunk(list, this.config.flushQueueSize);
-
-    await Promise.all(batches.map((batch) => this.send(batch, useRetry)));
+    this.filterTriableList();
+    if (this.queue.length === 0) {
+      return;
+    }
 
-    this.scheduleTryable(later);
+    const batches = chunk(this.queue, this.config.flushQueueSize);
+    for (const batch of batches) {
+      this.increaseAttempts(batch);
+      const responseStatus = await this.send(batch, useRetry);
+      // To keep the order of events sending to the backend, stop sending the other requests while error happending.
+      // All the events are still queued, and will be retried with next trigger.
+      if (retryStatus.has(responseStatus) && useRetry) {
```

Comment:
Would it make sense to check for success instead of checking for failure? There is only one success mode and many failure modes. If a new failure more is added later, if someone does not update this list, we'll end up with events being sent out of order.

Reply:
Why are we checking for retry when breaking here?

Reply:
> Why are we checking for retry when breaking here?
The previous logic is `useRetry = true` when SDK auto flush. And if the customer manually call `flush()`, `useRetry = false`. And the call will be directly `fulfill` without dealing with the response.
So here adding the `useRetry` to only break and resend later if useRetry = true.

Reply:
> Would it make sense to check for success instead of checking for failure? There is only one success mode and many failure modes. If a new failure more is added later, if someone does not update this list, we'll end up with events being sent out of order.

There are 2 more status we don't retry,   `Status.Failed` and `Status.Unknown`,  I think either way we need to have a list.

Reply:
There's an implicit dependency between the `send` function and this break statement here. Ideally, if the send function fulfilled the request, you can continue. If it did not fulfill it, then you cannot. Why can't that be returned and checked instead of statuses and useRetry?

For example, if the API key was missing, the send function will fulfill the request. Why should we break from the loop here and not send the rest? Similarly, if there was an unexpected error message, the items in the queue have been cleared. Why are we not moving forward?

Reply:
Order of requests of batches should be enforced if we send the next batch once the previous one is fulfilled regardless of the response and whether events in previous request are deleted or not.

Reply:
@Mercy811 not sure if you're agreeing with me or disagreeing :D fulfill == message deleted from storage right?

Reply:
Opinion 1:
> Would it make sense to check for success instead of checking for failure? 

Opinion 2:
> Ideally, if the send function fulfilled the request, you can continue. If it did not fulfill it, then you cannot. 

@izaaz I agree with your second opinion but not the first one

Reply:
Thanks @Mercy811. We are aligned. I had to look deeper to understand when we fulfill requests. So its not sufficient to just look at success vs failure.

Reply:
Thanks @izaaz @Mercy811 , 
Yes, the invalid status also includes unexpected errors and missing API, those shouldn't be retried.
It makes sense to continue if the request has been fulfilled. 
Fixed in the [last commit](https://github.com/amplitude/Amplitude-TypeScript/pull/736/commits/562735800bc7d54bc9e78b73864b7bd5e8425f56). Please check.